x/oracle/abci.go

```diff
@@ -17,77 +18,131 @@ func EndBlocker(ctx sdk.Context, k Keeper) {
 		return
 	}
 
-	actives := k.GetActiveDenoms(ctx)
-	votes := k.CollectVotes(ctx)
+	// Build valid votes counter and winner map over all validators in active set
+	validVotesCounterMap := make(map[string]int64)
+	winnerMap := make(map[string]types.Claim)
+	powerMap := make(map[string]int64)
+	k.StakingKeeper.IterateValidators(ctx, func(index int64, validator exported.ValidatorI) bool {
+
+		// Exclude not bonded vaildator or jailed validators from tallying
+		if validator.IsBonded() && !validator.IsJailed() {
+			valAddr := validator.GetOperator()
+			validVotesCounterMap[valAddr.String()] = int64(0)
+			winnerMap[valAddr.String()] = types.NewClaim(0, valAddr)
+			powerMap[valAddr.String()] = validator.GetConsensusPower()
+		}
 
-	// Clear prices
-	for _, activeDenom := range actives {
-		k.DeletePrice(ctx, activeDenom)
-	}
+		return false
+	})
 
-	// Changes whitelist array to map for the fast lookup
-	whitelistMap := make(map[string]bool)
+	whitelist := make(map[string]bool)
 	for _, denom := range k.Whitelist(ctx) {
-		whitelistMap[denom] = true
+		whitelist[denom] = true
 	}
 
-	// Iterate through votes and update prices; drop if not enough votes have been achieved.
-	claimMap := make(map[string]types.Claim)
-	for denom, ballot := range votes {
+	// Clear exchange rates
+	for denom := range whitelist {
+		k.DeleteLunaExchangeRate(ctx, denom)
+	}
+
+	// Organize votes to ballot by denom
+	voteMap := k.OrganizeBallotByDenom(ctx)
 
-		// Check whitelist; if denom is not exists or exists but the ballot is not passed, then skip
-		if _, exists := whitelistMap[denom]; !exists || !ballotIsPassing(ctx, ballot, k) {
+	// Build abstain validator map
+	abstainOperatorMap := make(map[string]bool)
+	abstainBallot := voteMap[core.MicroLunaDenom]
```

Comment:
Is the logic such that if a validator submits a vote for `uluna` he abstains from the entire ballot? 

It seems to be more elegant for him to abstain if the vote value is a negative number. For example, if the validator was submitting ballots for `krw, usd, sdr` and he chooses to abstain from `krw`, then he would submit: krw -1, usd 10, sdr 9... or something along this schema.

Reply:
agree